### PR TITLE
[zh] Sync blog 2023-05-05-memory-qos-cgroups-v2

### PR DESCRIPTION
**Pull request by wilsonwu**
Sync blog 2023-05-05-memory-qos-cgroups-v2:
```
content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md
```

### CODE REVIEW

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+<sub>for every j<sup>th</sup> container in every i<sup>th</sup> pod on a node</sub>
+-->
+<sub>对于节点中每个 i<sup>th</sup> Pod 中的每个 j<sup>th</sup> 容器</sub>
+<br>
+<br>
+<!--
+Kubelet will manage the cgroups hierarchy of the pod level and node level cgroups directly using the libcontainer library (from the runc project), while container cgroups limits are managed by the container runtime.
+-->
+Kubelet 将直接使用 libcontainer 库（来自 runc 项目）管理 Pod 级别和节点级别
+cgroup 的层次结构，而容器 cgroup 限制由容器运行时管理。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:344]:
```suggestion
cgroups 的层次结构，而容器 cgroups 限制由容器运行时管理。
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+### Support for Pod QoS classes
+### 支持 Pod QoS 类别 {#support-for-pod-qos-classes}
+
+<!--
+Based on user feedback for the Alpha feature in Kubernetes v1.22, some users would like to opt out of MemoryQoS on a per-pod basis to ensure there is no early memory throttling. Therefore, in Kubernetes v1.27 Memory QOS also supports memory.high to be set as per Quality of Service(QoS) for Pod classes. Following are the different cases for memory.high as per QOS classes:
+-->
+根据用户对 Kubernetes v1.22 中 Alpha 功能的反馈，一些用户希望选择退出基于每个 Pod 的 MemoryQoS，
+以确保不会出现早期内存限制。因此，在 Kubernetes v1.27 中 Memory QOS 还支持根据
+Pod 类的服务质量（QoS）设置 memory.high。以下是根据 QoS 类别的 memory.high 的不同情况：
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:354]:
```suggestion
服务质量（QoS）对 Pod 类设置 memory.high。以下是按 QoS 类设置 memory.high 的几种情况：
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+
+<!--
+1. **Guaranteed pods** by their QoS definition require memory requests=memory limits and are not overcommitted. Hence MemoryQoS feature is disabled on those pods by not setting memory.high. This ensures that Guaranteed pods can fully use their memory requests up to their set limit, and not hit any throttling.
+-->
+1. 根据 QoS 定义保证 Pod 需要内存请求 = 内存限制，并且不会过度使用。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:359]:
```suggestion
1. **Guaranteed Pods**：根据其 QoS 定义，要求 Pod 的内存请求等于其内存限制，并且不允许超配。
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+因此，通过不设置 memory.high，MemoryQoS 功能会在这些 Pod 上被禁用。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:360]:
```suggestion
   因此，通过不设置 memory.high，MemoryQoS 特性会针对这些 Pod 被禁用。
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+这确保了 **Guaranteed Pod** 可以充分利用其内存请求，达到设定的限制，并且不会遇到任何限制。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:361]:
```suggestion
   这样做可以确保 **Guaranteed Pod** 充分利用其内存请求，也就是其内存限制，并且不会被抑制。
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+
+<!--
+2. **Burstable pods** by their QoS definition require at least one container in the Pod with CPU or memory request or limit set.
+-->
+2. 根据 QoS 定义，**Burstable Pod** 要求 Pod 中至少有一个容器具有 CPU 或内存请求或限制设置。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:366]:
```suggestion
2. **Burstable Pod**：根据其 QoS 定义，要求 Pod 中至少有一个容器具有 CPU 或内存请求或限制设置。
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+
+   <!--
+   * When requests.memory and limits.memory are set, the formula is used as-is:
+   -->
+   * 当 requests.memory 和 limits.memory 设置时，公式按原样使用：
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:371]:
```suggestion
   * 当 requests.memory 和 limits.memory 都被设置时，公式按原样使用：
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+
+     <!--
+     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-limit.svg" title="memory.high when requests and limits are set" alt="memory.high when requests and limits are set" >}}
+     -->
+     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-limit.svg" title="当请求和限制被设置时的 memory.high" alt="当请求和限制被设置时的 memory.high" >}}
+
+   <!--
+   * When requests.memory is set and limits.memory is not set, limits.memory is substituted for node allocatable memory in the formula:
+   -->
+   * 当设置了 requests.memory 且未设置 limits.memory 时，公式中的 limits.memory 替换为节点可分配内存：
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:381]:
```suggestion
   * 当设置了 requests.memory 但未设置 limits.memory 时，公式中的 limits.memory 替换为节点可分配内存：
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+
+     <!--
+     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-no-limits.svg" title="memory.high when requests and limits are not set" alt="memory.high when requests and limits are not set" >}}
+     -->
+     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-no-limits.svg" title="当请求和限制未被设置时的 memory.high" alt="当请求和限制未被设置时的 memory.high" >}}
+
+<!--
+3. **BestEffort** by their QoS definition do not require any memory or CPU limits or requests. For this case, kubernetes sets requests.memory = 0 and substitute limits.memory for node allocatable memory in the formula:
+-->
+3. 根据 QoS 定义，**BestEffort** 不需要任何内存或 CPU 限制或请求。对于这种情况，
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:391]:
```suggestion
3. **BestEffort Pods**：根据其 QoS 定义，不需要设置内存或 CPU 限制或请求。对于这种情况，
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+   kubernetes 设置 requests.memory = 0 并用 Limits.memory 替换公式中的节点可分配内存：
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:392]:
```suggestion
   kubernetes 设置 requests.memory = 0 并将公式中的 limits.memory 替换为节点可分配内存：
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+
+<!--
+## How do I use it?
+-->
+## 我该如何使用它？ {#how-do-i-use-it}
+
+<!--
+The prerequisites for enabling Memory QoS feature on your Linux node are:
+-->
+在 Linux 节点上启用内存 QoS 功能的先决条件是：
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:413]:
```suggestion
在 Linux 节点上启用 MemoryQoS 特性的先决条件是：
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+
+2. 确保 CRI 运行时支持内存 QoS。在撰写本文时，
+   只有 Containerd 和 CRI-O 提供与内存 QoS（alpha）兼容的支持。是在以下 PR 中实现的：
+    * Containerd：[Feature: containerd-cri support LinuxContainerResources.Unified #5627](https://github.com/containerd/containerd/pull/5627)。
+    * CRI-O：[implement kube alpha features for 1.22 #5207](https://github.com/cri-o/cri-o/pull/5207)。
+
+<!--
+Memory QoS remains an alpha feature for Kubernetes v1.27. You can enable the feature by setting `MemoryQoS=true` in the kubelet configuration file:
+-->
+内存 QoS 仍然是 Kubernetes v1.27 的 Alpha 功能。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:432]:
```suggestion
MemoryQoS 在 Kubernetes v1.27 中仍然是 Alpha 特性。
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+你可以通过在 kubelet 配置文件中设置 `MemoryQoS=true` 来启用该功能：
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:433]:
```suggestion
你可以通过在 kubelet 配置文件中设置 `MemoryQoS=true` 来启用该特性：
```

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+        cpu: "500m"
+```
+
+* `spec.containers[].resources.requests`
+
+  <!--
+  When you specify the resource request for containers in a Pod, the [Kubernetes scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/#kube-scheduler) uses this information to decide which node to place the Pod on. The scheduler ensures that for each resource type, the sum of the resource requests of the scheduled containers is less than the total allocatable resources on the node.
+  -->
+  当你为 Pod 中的容器指定资源请求时，
+  [Kubernetes 调度程序](/zh-cn/docs/concepts/scheduling-eviction/kube-scheduler/#kube-scheduler)使用此信息来决定将 Pod 放置在哪个节点上。
```
[the remark this answers is not in the document]
Comment (wilsonwu) [reply content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md]: Done

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+<!--
+### Cgroups v2 memory controller interfaces & Kubernetes container resources mapping
+-->
+### Cgroups v2 内存控制器接口和 Kubernetes 容器资源映 {#cgroups-v2-memory-controller-interfaces-kubernetes-container-resources-mapping}
+
+<!--
+Memory QoS uses the memory controller of cgroups v2 to guarantee memory resources in Kubernetes. cgroupv2 interfaces that this feature uses are:
+-->
+内存 QoS 使用 cgroups v2 的内存控制器来保证 Kubernetes 中的内存资源。
+此功能使用的 cgroupv2 接口有：
```
[the remark this answers is not in the document]
Comment (wilsonwu) [reply content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md]: All fixed.

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 16]
```diff
@@ -0,0 +1,470 @@
...
+
+<!--
+{{< figure src="/blog/2023/05/05/qos-memory-resources/memory-qos-cal.svg" title="Memory QoS Levels" alt="Memory QoS Levels" >}}
+-->
+{{< figure src="/blog/2023/05/05/qos-memory-resources/memory-qos-cal.svg" title="内存 QoS 级别" alt="内存 QoS 级别" >}}
+
+<!--
+`memory.max` is mapped to `limits.memory` specified in the Pod spec. The kubelet and the container runtime configure the limit in the respective cgroup. The kernel enforces the limit to prevent the container from using more than the configured resource limit. If a process in a container tries to consume more than the specified limit, kernel terminates a process(es) with an out of memory Out of Memory (OOM) error.
+-->
+`memory.max` 映射到 Pod 规范中指定的 `limits.memory`。
```
[the remark this answers is not in the document]
Comment (wilsonwu) [reply content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md]: All fixed.

File: content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md [hunk 17]
```diff
@@ -0,0 +1,470 @@
...
+<sub>对于节点中每个 i<sup>th</sup> Pod 中的每个 j<sup>th</sup> 容器</sub>
+<br>
+<br>
+<!--
+Kubelet will manage the cgroups hierarchy of the pod level and node level cgroups directly using the libcontainer library (from the runc project), while container cgroups limits are managed by the container runtime.
+-->
+Kubelet 将直接使用 libcontainer 库（来自 runc 项目）管理 Pod 级别和节点级别
+cgroup 的层次结构，而容器 cgroup 限制由容器运行时管理。
+
+### Support for Pod QoS classes
```
[the remark this answers is not in the document]
Comment (wilsonwu) [reply content/zh-cn/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md:347]: Fixed.

**Review by ydFu — APPROVED** [review]
/approve

**Pull request merged**